audio/data.py

```diff
@@ -258,66 +262,86 @@ def __init__(self, items, path, config=AudioConfig(), **kwargs):
         if str(path) not in str(cd): 
             config.cache_dir = path / cd
         self.config = config
+        self.nchannels = 1 if self.config.downmix else None
         self.copy_new += ['config']
 
-    def open(self, item) -> AudioItem:
-        p = Path(item)
-        if self.path is not None and not p.exists() and str(self.path) not in str(item): p = self.path/item
-        if not p.exists(): raise FileNotFoundError(f"Neither '{item}' nor '{p}' could be found")
-        if not str(p).lower().endswith(AUDIO_EXTENSIONS): raise Exception("Invalid audio file")
-
-        cfg = self.config
-        if cfg.use_spectro:
-            folder = md5(str(asdict(cfg))+str(asdict(cfg.sg_cfg)))
-            fname = f"{md5(str(p))}-{p.name}.pt"
-            image_path = cfg.cache_dir/(f"{folder}/{fname}")
-            if cfg.cache and not cfg.force_cache and image_path.exists():
-                mel = torch.load(image_path)
-                start, end = None, None
-                if cfg.duration and cfg._processed:
-                    mel, start, end = tfm_crop_time(mel, cfg._sr, cfg.duration, cfg.sg_cfg.hop, cfg.pad_mode)
-                return AudioItem(spectro=mel, path=item, max_to_pad=cfg.max_to_pad, start=start, end=end)
-
-        sig, sr = torchaudio.load(str(p))
-        if(cfg._sr is not None and sr != cfg._sr):
-            raise ValueError(f'''Multiple sample rates detected. Sample rate {sr} of file {str(p)} 
-                                does not match config sample rate {cfg._sr} 
+    def open(self, fn): # file name, it seems
+        item = open_audio(fn)
```

Comment:
You've lost the lazy loading here. 

Previously we would only open the audio file if the cache wasn't found and the use_spectro was true. The code was a bit complex but it was faster because of this. Could you add that back in?

Reply:
Can confirm. When we are training on spectrograms there's no reason to load the audio as it takes time and we don't use it. If the spectrogram is coming from the cache, we only want to load the audio when the user requests to see it via `.show()`.  The easiest fix that preserves the new pipeline and improved modularity would be to...

In `open_audio()` check first to see if they are using spectros, and if the spectro is already cached. If so, grab it and return a new `AudioItem` with that spectro and sig/sr = None. If there isn't a spectro, load the audio file and return an `AudioItem` containing sig/sr but spectro=None (or if `use_spectro` = True you could call create_spectro and generate it right there). 

Doing this means that you'll have to go through the rest of the pipeline and see what is affected and fix it. But it seems like the best way to preserve the structure you've set up by generating an `AudioItem` up front and handing it off to each function for validation/spectro generation...etc